llvm/lib/CodeGen/MachineScheduler.cpp

```diff
@@ -4266,6 +4266,12 @@ unsigned ResourceSegments::getFirstAvailableAt(
   assert(std::is_sorted(std::begin(_Intervals), std::end(_Intervals),
                         sortIntervals) &&
          "Cannot execute on an un-sorted set of intervals.");
+
+  // Zero resource usage is allowed by TargetSchedule.td but we do not construct
+  // a ResourceSegment interval for that situation.
+  if (AcquireAtCycle == Cycle)
+    return Cycle;
```

Comment:
This seems to be wrong. Say that `CurrCycle` is 200, and that ResourceX have a `[AcquireAtCycle,ReleaseAtCycle]` pair `[5,5]` (note that `Cycle == ReleaseAtCycle` in the naming of the method's parameters... unfortunate)

Your code will say that the resource is available at cycle 5 (not even considering if the resource is busy at cycle 200), but in reality we can only schedule from cycle 200 onwards.

Reply:
I have updated to return CurrCycle, which would be 200 in this case. It should return CurrCycle since it is as if the for loop below hit the continue statement every time since nothing intersects with the empty interval.

Reply:
I will update Cycle -> ReleaseAtCycle in a future patch.